File: android/src/main/kotlin/io/agora/agorartm/Extensions.kt
package io.agora.agorartm

import io.agora.rtm.ChannelAttributeOptions
import io.agora.rtm.ErrorInfo
import io.agora.rtm.LocalInvitation
import io.agora.rtm.RemoteInvitation
import io.agora.rtm.RtmAttribute
import io.agora.rtm.RtmChannelAttribute
import io.agora.rtm.RtmChannelMember
import io.agora.rtm.RtmChannelMemberCount
import io.agora.rtm.RtmClient
import io.agora.rtm.RtmMessage
import io.agora.rtm.RtmServiceContext
import io.agora.rtm.SendMessageOptions

fun RtmMessage.toJson(): Map<String, Any?> {
    return hashMapOf(
        "text" to text,
        "rawMessage" to rawMessage,
        "messageType" to messageType,
        "serverReceivedTs" to serverReceivedTs,
        "isOfflineMessage" to isOfflineMessage,
    )
}

fun Map<*, *>.toRtmMessage(client: RtmClient?): RtmMessage? {
    val text = this["text"] as? String
    return (this["rawMessage"] as? ByteArray)?.let {
        client?.createMessage(it, text)
    } ?: let { client?.createMessage(text) }
}

fun RtmChannelMember.toJson(): Map<String, Any?> {
    return hashMapOf(
        "userId" to userId,
        "channelId" to channelId,
    )
}

fun RtmAttribute.toJson(): Map<String, Any?> {
    return hashMapOf(
        "key" to key,
        "value" to value,
    )
}

fun Map<*, *>.toRtmAttribute(): RtmAttribute {
    return RtmAttribute().apply {
        key = this@toRtmAttribute["key"] as? String
        value = this@toRtmAttribute["value"] as? String
    }
}

fun List<*>.toRtmAttributeList(): List<RtmAttribute?> {
    return this.map { (it as? Map<*, *>)?.toRtmAttribute() }
}

fun RtmChannelAttribute.toJson(): Map<String, Any?> {
    return hashMapOf(
        "key" to key,
        "value" to value,
        "lastUpdateUserId" to lastUpdateUserId,
        "lastUpdateTs" to lastUpdateTs,
    )
}

fun Map<*, *>.toRtmChannelAttribute(): RtmChannelAttribute {
    return RtmChannelAttribute().apply {
        key = this@toRtmChannelAttribute["key"] as? String
        value = this@toRtmChannelAttribute["value"] as? String
    }
}

fun List<*>.toRtmChannelAttributeList(): List<RtmChannelAttribute?> {
    return this.map { (it as? Map<*, *>)?.toRtmChannelAttribute() }
}

fun LocalInvitation.toJson(): Map<String, Any?> {
    return hashMapOf(
        "calleeId" to calleeId,
        "content" to content,
        "channelId" to channelId,
        "response" to response,
        "state" to state,
        "hashCode" to hashCode(),
    )
}

fun Map<*, *>.toLocalInvitation(callManager: RTMCallManager): LocalInvitation {
    val hashCode = this["hashCode"] as? Int
    return hashCode?.let {
        callManager.localInvitations[it]?.apply {
            content = this@toLocalInvitation["content"] as? String
            channelId = this@toLocalInvitation["channelId"] as? String
        }
    } ?: let {
        val calleeId = it["calleeId"] as? String
        callManager.manager.createLocalInvitation(calleeId).apply {
            content = it["content"] as? String
            channelId = it["channelId"] as? String
        }
    }
}

fun RemoteInvitation.toJson(): Map<String, Any?> {
    return hashMapOf(
        "callerId" to callerId,
        "content" to content,
        "channelId" to channelId,
        "response" to response,
        "state" to state,
        "hashCode" to hashCode(),
    )
}

fun Map<*, *>.toRemoteInvitation(callManager: RTMCallManager): RemoteInvitation? {
    val hashCode = this["hashCode"] as? Int
    return callManager.remoteInvitations[hashCode]?.apply {
        response = this@toRemoteInvitation["response"] as? String
    }
}

fun ChannelAttributeOptions.toJson(): Map<String, Any?> {
    return hashMapOf(
        "enableNotificationToChannelMembers" to enableNotificationToChannelMembers,
    )
}

fun Map<*, *>.toChannelAttributeOptions(): ChannelAttributeOptions {
    return ChannelAttributeOptions().apply {
        enableNotificationToChannelMembers =
            this@toChannelAttributeOptions["enableNotificationToChannelMembers"] as Boolean
    }
}

fun SendMessageOptions.toJson(): Map<String, Any?> {
    return hashMapOf(
    )
}

fun Map<*, *>.toSendMessageOptions(): SendMessageOptions {
    return SendMessageOptions()
}

fun RtmChannelMemberCount.toJson(): Map<String, Any?> {
    return hashMapOf(
        "channelId" to channelID,
        "memberCount" to memberCount,
    )
}

fun ErrorInfo.toJson(): Map<String, Any?> {
    return hashMapOf(
        "errorCode" to errorCode,
        "errorDescription" to errorDescription,
    )
}

fun Map<*, *>.toRtmServiceContext(): RtmServiceContext {
    return RtmServiceContext().apply {
        areaCode = this@toRtmServiceContext["areaCode"] as Int
        proxyType = this@toRtmServiceContext["proxyType"] as Int
    }
}

fun List<*>.toStringSet(): Set<String> {
    return toStringList().toSet()
}

fun List<*>.toStringList(): List<String> {
    return this.map { it as String }
}

fun List<*>.toJson(): List<Map<String, Any?>> {
    return this.map {
        when (it) {
            is RtmAttribute -> it.toJson()
            is RtmChannelAttribute -> it.toJson()
            is RtmChannelMember -> it.toJson()
            is RtmChannelMemberCount -> it.toJson()
            else -> emptyMap()
        }
    }
}

File: ios/Classes/Extensions.swift
import AgoraRtmKit
import Flutter

extension AgoraRtmMessage {
    func toJson() -> [String: Any?] {
        switch(self){
        case let self as AgoraRtmRawMessage: return [
            "text": text,
            "rawMessage": FlutterStandardTypedData(bytes: self.rawData),
            "messageType": type.rawValue,
            "serverReceivedTs": serverReceivedTs,
            "isOfflineMessage": isOfflineMessage,
        ]
        default: return [
            "text": text,
            "messageType": type.rawValue,
            "serverReceivedTs": serverReceivedTs,
            "isOfflineMessage": isOfflineMessage,
        ]
        }
    }
}

extension AgoraRtmMember {
    func toJson() -> [String: Any?] {
        return [
            "userId": userId,
            "channelId": channelId,
        ]
    }
}

extension AgoraRtmAttribute {
    func toJson() -> [String: Any?] {
        return [
            "key": key,
            "value": value,
        ]
    }
}

extension AgoraRtmChannelAttribute {
    func toJson() -> [String: Any?] {
        return [
            "key": key,
            "value": value,
            "lastUpdateUserId": lastUpdateUserId,
            "lastUpdateTs": lastUpdateTs,
        ]
    }
}

extension AgoraRtmLocalInvitation {
    func toJson() -> [String: Any?] {
        return [
            "calleeId": calleeId,
            "content": content,
            "channelId": channelId,
            "response": response,
            "state": state.rawValue,
            "hashCode": hash,
        ]
    }
}

extension AgoraRtmRemoteInvitation {
    func toJson() -> [String: Any?] {
        return [
            "callerId": callerId,
            "content": content,
            "channelId": channelId,
            "response": response,
            "state": state.rawValue,
            "hashCode": hash,
        ]
    }
}

extension AgoraRtmChannelAttributeOptions{
    func toJson() -> [String: Any?] {
        return [
            "enableNotificationToChannelMembers": enableNotificationToChannelMembers,
        ]
    }
}

extension AgoraRtmSendMessageOptions {
    func toJson() -> [String: Any?] {
        return [:]
    }
}

extension AgoraRtmChannelMemberCount {
    func toJson() -> [String: Any?] {
        return [
            "channelId": channelId,
            "memberCount": count,
        ]
    }
}

extension Dictionary where Key == String {
    func toRtmMessage() -> AgoraRtmMessage {
        let text = self["text"] as? String
        if let rawMessage = self["rawMessage"] as? FlutterStandardTypedData {
            return AgoraRtmRawMessage(rawData: rawMessage.data, description: text!)
        } else {
            return AgoraRtmMessage(text: text!)
        }
    }
    
    func toRtmAttribute() -> AgoraRtmAttribute {
        let rtmAttribute = AgoraRtmAttribute()
        if let key = self["key"] as? String {
            rtmAttribute.key = key
        }
        if let value = self["value"] as? String {
            rtmAttribute.value = value
        }
        return rtmAttribute
    }
    
    func toRtmChannelAttribute() -> AgoraRtmChannelAttribute {
        let rtmAttribute = AgoraRtmChannelAttribute()
        if let key = self["key"] as? String {
            rtmAttribute.key = key
        }
        if let value = self["value"] as? String {
            rtmAttribute.value = value
        }
        return rtmAttribute
    }
    
    func toLocalInvitation(_ callManager: RTMCallManager) -> AgoraRtmLocalInvitation? {
        if let hashCode = self["hashCode"] as? Int, let localInvitation = callManager.localInvitations[hashCode] {
            localInvitation.content = self["content"] as? String
            localInvitation.channelId = self["channelId"] as? String
            return localInvitation
        }
        if let calleeId = self["calleeId"] as? String {
            let localInvitation = AgoraRtmLocalInvitation(calleeId: calleeId)
            localInvitation.content = self["content"] as? String
            localInvitation.channelId = self["channelId"] as? String
            return localInvitation
        }
        return nil
    }
    
    func toRemoteInvitation(_ callManager: RTMCallManager) -> AgoraRtmRemoteInvitation? {
        if let hashCode = self["hashCode"] as? Int, let remoteInvitation = callManager.remoteInvitations[hashCode] {
            remoteInvitation.response = self["response"] as? String
            return remoteInvitation
        }
        return nil
    }
    
    func toChannelAttributeOptions() -> AgoraRtmChannelAttributeOptions {
        let channelAttributeOptions = AgoraRtmChannelAttributeOptions()
        if let enableNotificationToChannelMembers = self["enableNotificationToChannelMembers"] as? Bool {
            channelAttributeOptions.enableNotificationToChannelMembers = enableNotificationToChannelMembers
        }
        return channelAttributeOptions
    }
    
    func toSendMessageOptions() -> AgoraRtmSendMessageOptions {
        return AgoraRtmSendMessageOptions()
    }

    func toRtmServiceContext() -> AgoraRtmServiceContext {
        let context = AgoraRtmServiceContext()
        if let areaCode = self["areaCode"] as? UInt {
            context.areaCode = AgoraRtmAreaCode(rawValue: areaCode)
        }
        if let proxyType = self["proxyType"] as? UInt {
            context.proxyType = AgoraRtmCloudProxyType(rawValue: proxyType)
        }
        return context
    }
}

extension Array {
    func toRtmAttributeList() -> [AgoraRtmAttribute] {
        return self.map { ($0 as? [String: Any?] ?? [:]).toRtmAttribute() }
    }
    
    func toRtmChannelAttributeList() -> [AgoraRtmChannelAttribute] {
        return self.map { ($0 as? [String: Any?] ?? [:]).toRtmChannelAttribute() }
    }
    
    func toStringList() -> [String] {
        return self.map { $0 as? String ?? "" }
    }
    
    func toJson() -> [[String: Any?]] {
        return self.map {
            switch $0 {
            case let it as AgoraRtmAttribute: return it.toJson()
            case let it as AgoraRtmChannelAttribute: return it.toJson()
            case let it as AgoraRtmMember: return it.toJson()
            case let it as AgoraRtmChannelMemberCount: return it.toJson()
            default: return [:]
            }
        }
    }
}
